cirrus/google_cloud/utils.py

```diff
@@ -230,6 +236,126 @@ def _get_string_to_sign(
     return string_to_sign
 
 
+def get_signed_url(
+    path_to_resource,
+    http_verb,
+    expires,
+    extension_headers=None,
+    canonical_query_params=None,
+    service_account_creds=None,
+    requester_pays_user_project=None,
+):
+    """
+    V4 signing.
+
+    Requirements/process:
+        https://cloud.google.com/storage/docs/access-control/signing-urls-manually
+
+    Args:
+        path_to_resource (str): The path/url to the resources/google bucket. This is
+            everything that follows the host name but precedes any query strings.
+        http_verb (str): The HTTP verb. Includes DELETE, GET, HEAD, POST*, PUT.
+            Signed URLs do not support POST requests, except when working with
+            resumable uploads.
+        expires (int): The amount of time (in seconds) before the signed url will expire.
+        extension_headers (dict, optional): Optional request headers.
+        canonical_query_params(dict, optional): Optional query strings to add to the request
+        service_account_creds (dict, optional): JSON keyfile dict for Google
+            Service Account (can be obtained by calling `get_access_key`)
+        requester_pays_user_project (optional): User's Google project for billing
+
+    Returns:
+        str: Completed signed URL
+    """
+    service_account_creds = (
+        service_account_creds or config.GOOGLE_APPLICATION_CREDENTIALS
+    )
+    creds = service_account.Credentials.from_service_account_info(service_account_creds)
+
+    bucket_name = path_to_resource.split("/")[0]
+    object_name = "/".join(path_to_resource.split("/")[1:])
+    escaped_object_name = quote(object_name.encode(), safe=b"/~")
+    canonical_uri = "/{}".format(escaped_object_name)
+
+    datetime_now = datetime.datetime.utcnow()
+    request_timestamp = datetime_now.strftime("%Y%m%dT%H%M%SZ")
+    datestamp = datetime_now.strftime("%Y%m%d")
+
+    client_email = creds.service_account_email
+    credential_scope = "{}/auto/storage/goog4_request".format(datestamp)
+    credential = "{}/{}".format(client_email, credential_scope)
+
+    if extension_headers is None:
+        extension_headers = dict()
+
+    host = (
+        "{}.storage.googleapis.com".format(bucket_name)
+        if bucket_name
+        else "storage.googleapis.com"
+    )
```

Comment:
> in which case do we have no bucket name?
i think if we have no bucket name, you'd end up with an empty `object_name` on line 283 🤔 

you resolved my comment without answering. i should be the one resolving 🙂 
what's going on here?

Reply:
Oops sorry, yeah this is needed because for some reason bucket names aren't parsed.

Reply:
did this happen when you were testing? when does it happen? sounds like a bug

Reply:
yeah, when testing with the requestor pays that was set up the output url was ".storage.googleapis.com" without this conditional. It could be that the path, instead of bucket/object it is just bucket. I will test this.